### ISSUE TITLE
[feat] Lift max_batch_size=1 under --ep-size 2 by multiplexing the head↔worker protocol

### ISSUE THREAD

**Issue by camerono** [Avarok-Cybersecurity/atlas#99]
### Motivation

When Atlas runs with `--ep-size 2 --world-size 2`, the server hard-pins `max_batch_size = 1` regardless of what the operator passes. That's spelled out in `crates/spark-server/src/main_modules/serve.rs:307-314`:

```rust
// EP: force batch_size=1 (worker protocol is single-sequence).
// MTP speculative decoding IS supported with EP via verify broadcast protocol.
let max_batch_size = if world_size > 1 {
    tracing::info!("EP active: forcing max_batch_size=1");
    1
} else {
    args.max_batch_size
};
```

For Qwen3.5-122B-A10B-NVFP4 on 2× GB10 this gate becomes the structural concurrency ceiling, and the model doesn't have alternative-topology relief. Tried both single-node Atlas and `--tp-size 2` on the same hardware. Both load all ~94 GB of weights onto the head and the KV cache pool gets 0 allocatable blocks even at `--max-batch-size 1` + `--max-seq-len 16384` + `--gpu-memory-utilization 0.95`. EP=2 is the only topology that actually shards (expert-parallel), so the EP-gate ends up being load-bearing.

What it costs in practice: a 4-concurrent agent-style burst (multi-tool agent-style Dify workflow doing parallel tool dispatch) puts 4 chat completions in flight against Atlas, the scheduler queues them serially behind the batch=1 cap, and the upstream client (Dify's plugin daemon) hits its internal 500/retry path after ~10 minutes per slot. 2/4 requests in my Bench 3 k=4 tail-spike to ~605 s with `Request to Plugin Daemon Service failed-500`. With batch=4 (or even 2) on the same Atlas process this would be a non-issue.

### Proposal

Extend the head↔worker RPC to multiplex sequences so the EP-active branch can honor `args.max_batch_size`. Concretely:

1. Add a `seq_id: u32` (or similar) to the head→worker request payload.
2. Have the worker allocate `max_batch_size` parallel KV cache slices + SSM state pool slots, parallel to what the head already does at `world_size == 1`.
3. Tag returned tokens with their `seq_id` so the head routes back to the right scheduler entry.
4. Audit the MTP `verify_broadcast` path for multi-sequence correctness — the existing single-sequence guarantee is what made that broadcast tractable, and the multi-seq case is the gnarliest interaction here.
5. Flip the `world_size > 1` check to honor `args.max_batch_size`.

The scheduler side is already designed for batching — `crates/spark-server/src/scheduler/mod.rs` uses `max_batch_size` consistently in the active-set caps, prefill drain logic, etc. The constraint is purely the EP worker protocol, not the scheduler.

Happy to attempt the PR if you've not already started — would want maintainer sign-off on the protocol shape before spending the weeks on this. The speculative-decoding audit is the unknown-unknown.

### Scope

Server / API change

### Alternatives considered

- `--tp-size 2`: no per-node weight relief on this model (head still loads ~94 GB), so single-node-style OOM. The `--tp-size` flag is parsed by `cli.rs` but for Qwen3.5-122B-A10B it doesn't appear to actually shard weights.
- Single-node EP=0: weights consume 94 GB of the 121 GiB unified pool, leaves Atlas with 0 allocatable KV blocks even at batch=1.
- Smaller model: would defeat the quality lift Qwen3.5-122B got us in our agent eval (6 pp pass@1 over Qwen3-Coder-Next-80B in code-bench's earlier polyglot run).
- Front-side queue limiter on the Dify side: lands today and makes failure polite, but doesn't lift the cap. That's our near-term mitigation regardless of whether this lands.

### Additional context

- Code location: `crates/spark-server/src/main_modules/serve.rs:307-314`.
- Comment in source: *"worker protocol is single-sequence"* — what made me read this as defensive engineering rather than a fundamental limit.
- Hardware: 2× GB10, ROCE+InfiniBand interlink, NCCL via the 200G QSFP pair.
- Image we hit this on: `avarok/atlas-gb10@sha256:53f5c6df78a525d64e4b8757a97a973279fda74eb080483c0cc1bb0d05545536` (current stacked-best config: `--speculative --mtp-quantization nvfp4 --disable-thinking --enable-prefix-caching true --default-min-p 0.05 --kv-high-precision-layers auto`).
- Atlas issue #94 we filed separately (`--ngram-speculative` / `--self-speculative` broken on this model) is a sibling but independent. This one is about the EP gate; that one is about the alternative spec paths failing.

—

I'm grateful for this project's efforts and appreciate the chance to help evolve/improve it.

**Comment by camerono** [comment]
Spent a session reading the code with this issue in mind. The change is smaller than I framed in the issue body — most of the multi-slot machinery is already there. Posting the design before writing Rust to get a sanity check from the project before I burn cycles in the wrong direction.

## Key finding

`SequenceState.slot_idx: usize` (`crates/spark-model/src/traits.rs:86`) already identifies the SSM-pool slot for a sequence. The whole forward-pass machinery indexes by it — CUDA graph caches in `decode_a.rs`, `verify_b.rs`, `verify_c.rs`, `verify_c2.rs` are `HashMap<slot_idx, GraphHandle>`. SSM snapshot pool keys by it. Async checkpoint slots key by it. `SsmStatePool` pre-allocates `max_batch_size + 1` slots with a `Mutex<Vec<usize>>` free-list and `claim_slot()` / `release_slot(idx)` (`ssm_pool.rs:151-159`).

The single thing the EP protocol can't express is *which* slot the head is operating on. So the worker's `loop { ep_worker_step(&mut seq) }` only ever uses slot 0. That's why `serve.rs:307-314` is forced to clamp `max_batch_size = 1` — anything higher would have the head and worker disagreeing about which slot's state is being read/written.

`slot_idx` is the seq_id. We don't need to add a parallel identifier — just plumb the existing one through the protocol.

## Proposed protocol shape

Add a `seq_id` u32 broadcast immediately before each command broadcast. The head sends `(slot_idx as u32) → existing_cmd_code → existing_follow_on_broadcasts`. The worker reads the preamble first and routes the rest of the command's data into `slots[seq_id]` rather than the singleton `seq`. Existing command codes (`0xFFFFFFF0` prefill, `0xFFFFFFF2/3/4` verify K=2/3/4, decode token IDs in `0..0xFFFFFFEF`, etc.) keep their semantics — they just become slot-bound.

Behind a `ATLAS_EP_PROTOCOL=v2` flag for the first few commits so the wire change is opt-in during development. Flag can be retired at merge if you'd prefer the change be straight backward-incompatible (single-image releases make this safe).

## Five-commit landing plan

1. **Protocol primitives.** `ep_broadcast_seq_and_cmd(seq_id, cmd)` + `ep_recv_seq_and_cmd() -> (u32, u32)` helpers in `impl_a2.rs`. Wraps the two broadcasts; when v2 flag is off, the preamble is skipped and behaviour matches today. No callers updated yet. Unit-test the roundtrip.
2. **Head emits the preamble.** Every call site of `ep_broadcast_cmd` in `crates/spark-server/src/scheduler/` (decode loop, prefill, `verify_k{2,3,4}_step.rs`) gets the preamble inserted, using `a.seq.slot_idx as u32` as the seq_id. Falls back when v2 off.
3. **Worker dispatches multi-slot.** `build.rs:211` worker loop replaces the singleton `let mut seq = model.alloc_sequence()?` with `let mut slots: Vec<Option<SequenceState>> = (0..max_batch_size).map(|_| None).collect()`. `ep_worker_step_impl` signature changes from `(&self, seq: &mut SequenceState)` to `(&self, slots: &mut [Option<SequenceState>])`; reads `(seq_id, cmd)` via the helper, dispatches alloc commands to `slots[seq_id] = Some(self.alloc_sequence()?)`, all other commands route to `slots[seq_id].as_mut()` for the existing per-command logic.
4. **Alloc/free protocol cleanup.** Today's `0xFFFFFFF1` is "free-and-realloc the singleton" — replace with explicit `0xFFFFFFF1 = alloc-slot` (creates new sequence at the seq_id) and a new `free-slot` code for explicit slot release. The preamble determines which slot.
5. **Flip the gate.** Drop the `world_size > 1` clamp in `serve.rs:307-314`. Scheduler picks the next sequence to dispatch by round-robin over `active` rather than always-position-0 — smallest change to make multi-slot meaningful on the head side; smarter dispatch policy can be a follow-up. GPU integration test on a 2-node Spark with `--ep-size 2 --max-batch-size 4` against single-sequence reference output. MTP cross-check (accept rates against `--ep-size 2 --max-batch-size 1` baseline) is the critical correctness gate.

## What this PR does NOT do

It's option B (multiplexed protocol, sequential forward pass) — N sequences make progress concurrently from the user's POV, but each worker forward pass still operates on one slot at a time. Time-sliced, not GPU-batched. The user-visible 605s tail (issue body's motivating example) disappears, but per-user latency is roughly N× slower than a single-user run since they share GPU time round-robin. The "true batched EP" version (kernels operate on N slots in one forward pass) is option A and would be a follow-up PR — same protocol foundation, much bigger surface area inside the worker forward pass.

## Open questions before I start

1. **The seq_id-equals-slot_idx identification.** I'm relying on head and worker both calling `ssm_pool.claim_slot()` / `release_slot()` in matched order so their independently-allocated slots stay aligned. Defensible (today's single-sequence behaviour relies on the same matched-order invariant), but worth confirming you haven't found a case where the order can drift.
2. **`ATLAS_EP_PROTOCOL=v2` flag lifetime.** Permanent for backward compat, or remove at merge?
3. **Commits 3+4 as separate commits vs one PR.** AGENTS.md says one logical change per commit, which I read as commits-are-cheap. CONTRIBUTING.md "one logical PR" reads broader. Happy either way — let me know your preference.

I'd appreciate a quick sanity check on the seq_id/slot_idx identification before I write the first commit. Going to start the protocol-primitive commit on a separate branch in my fork (`feat/batched-ep-worker-protocol`) so it's reviewable in isolation regardless of the answer.

**Comment by tbraun96** [comment]
Thanks for this! I think that having everything in one PR, in this case, makes sense. I would also encourage implementing the "true batched EP" you spoke of. The fewer shortcuts in our codebase, the more it will pay off long-term. Feel free to start a PR; we do not have the bandwidth right now to handle this, but can certainly merge in any work if it checks out.

**Comment by camerono** [comment]
Opened #101 with the implementation. The five-commit plan landed essentially as written — `slot_idx` as the seq_id, worker multi-slot dispatch, the `v2` flag. Three things diverged from what I expected above, and I measured the gains rather than assuming them.

**1. The protocol was the easy part; the head-side batched decode was the hard part.** The wire multiplexing went in cleanly. What bit me was `decode_batch_dispatch` on the head: its EP branch wrote single-row logits to row 0 per sequence, so every sequence sampled the last one's logits; and broadcasting all the seq-id preambles up front put the head's comm-stream op order out of step with the worker's per-layer all-reduces, which deadlocked NCCL. Both were in the model layer, not the protocol — fixed by interleaving the broadcasts inside the dispatch and staging each sequence's logits row.

**2. Lifting the gate fixes the tail latency — but aggregate throughput doesn't scale at low concurrency.** The k=4 tail-spike this issue is about is gone: the 4 requests now batch and finish together instead of queuing serially behind batch=1. But batched vs serialized aggregate throughput at N=2 is ~equal (~32 tok/s either way). On 2× GB10 the 122B decode is memory-bandwidth + inter-node-NCCL bound, and two concurrent tokens share almost no weight loads (the SSM projections are sequential GEMVs and the MoE experts are mostly disjoint across the 256-expert pool at low N). So the gate-lift buys admission and tail latency — exactly what I filed — not tok/s scaling. Worth setting that expectation.

**3. The batched MoE wants the fused batch-2/3 kernels, not the grouped-GEMM.** The SSM layers' batched MoE was dead code behind an early return; reviving it with `forward_prefill` (the prefill grouped-GEMM) was a ~60% regression — at per-expert M≈1 the 256-expert sort/permute overhead dominates, ×36 SSM layers. `forward_k2`/`forward_k3` for N=2/3 and a per-token loop for N≥4 instead.

Material gains measured (Qwen3.5-122B-A10B-NVFP4, EP=2, 2× GB10, MTP on):

| Change | Effect |
| --- | --- |
| Gate lift | k=4 burst batches instead of serializing — removes the batch=1 tail-spike from Bench 3 |
| SSM MoE dispatch (`forward_k2`) | SSM decode step 44 ms → 35 ms at N=2 (~15–20%) |
| Attention MoE dispatch | attention block ~40 ms → ~24 ms at N=4 |
| Aggregate at N=2 | ~32 tok/s (batched ≈ serialized — tail-latency win, not throughput) |

CUDA graphs and single-host were both dead ends — graphs give no win because the step isn't launch-bound, and the model doesn't fit one GB10 (KV pool gets 0 blocks), which reconfirms EP=2 is the only viable topology. The reasoning is in ADR-0011 in the PR.

**Comment by camerono** [comment]
One more, for anyone with a **3rd GB10** who wants to extend this: the test I couldn't run is EP=3, and two things make it worth doing.

First, decode here is memory-bandwidth bound and EP shards the experts, so a third rank cuts the expert bytes each rank streams per step. That makes it the clean way to find out whether decode throughput actually *scales* with EP degree — the question the 2-rank numbers can't answer (where batching only bought tail latency).

Second, it exercises the protocol past the 2-rank fast path. `all_reduce_2rank` — the `ncclSend`/`ncclRecv` + local-add shortcut — is gated on `world_size == 2`, so EP=3 falls back to the general `ncclAllReduce` and would shake out the batched-decode and MoE-dispatch paths under a real N-rank collective. The extra sharding also frees per-rank memory to push `--max-batch-size` or context length further.

On a personal note — this was genuinely fun. Most of my work in technology over the last 20 years has evolved towards almost exclusive administrative-side focus with the occasional dive 'into the depths'; this was a chance to get back to my roots.

**Comment by tbraun96** [comment]
@camerono thanks for both the issue and the follow-up design comment, the second one especially. Reading the code and finding that slot_idx already is the seq_id, so the protocol just could not express it, is exactly the right diagnosis and it kept the change small. #101 implements this and I reviewed it in depth over there. Static verification is done (builds, both test suites green, v1 path unchanged behind ATLAS_EP_PROTOCOL=v2). I am queuing a live 2x GB10 EP=2 run on Qwen3.5-122B-A10B-NVFP4 to confirm multi-seq isolation and the MTP verify path before it lands. Keeping this open until that merges.

**Comment by tbraun96** [comment]
Landed in #101 and validated live on 2x GB10 EP=2 with Qwen3.5-122B-A10B-NVFP4. @camerono the slot-mux protocol works: ATLAS_EP_PROTOCOL=v2 honors --max-batch-size (saw max_batch=4 instead of the old forced 1), concurrent multi-sequence decode is correctly isolated at N=2/4/8 with no cross-sequence contamination, NCCL inter-node bring-up is clean, and both nodes stayed stable through the burst. Your original read of the issue, that slot_idx already was the seq_id and the protocol just could not express it, was exactly right and kept the change tight. Closing this as resolved. Thanks for the issue and the design comment that drove it.
